### ISSUE TITLE
Dreaming — idle-time memory consolidation loop (retro emits, dreaming promotes)

### ISSUE THREAD

**Issue by souliane** [souliane/teatree#1933]
# Dreaming — idle-time memory consolidation loop

## 1. Motivation

The agent's long-term memory is a directory of markdown files plus an always-loaded index (`MEMORY.md`). It grows monotonically and nothing prunes or compacts it.

Concrete state today:
- `MEMORY.md` is **520 lines / 116.3KB**, and the loader now truncates it: *"WARNING: MEMORY.md is 520 lines and 116.3KB. Only part of it was loaded. Keep index entries to one line under ~200 chars; move detail into topic files."* When the index overflows the load budget, entries silently stop being seen — the memory exists on disk but no longer influences behavior.
- The memory directory holds **~605 topic files** (`feedback_*.md` / `project_*.md` / `reference_*.md`).
- The index header already admits drift: *"Many older 'obvious-now' feedback entries are still on disk but unindexed; restore here if a recurrence is observed."* That is an informal cold-storage rule that no process enforces.
- Entries accumulate near-duplicate feedback rules (multiple corrections that share one root cause), stale `project_*` snapshots for shipped/abandoned work, and BINDING rules buried below the truncation line.

The only existing memory code on the teatree side (`src/teatree/memory_audit.py`) is a **read-only scanner** — it flags guardrail-shaped entries that should graduate to skills. There is no dedup, no merge, no re-index, no archive. Index bloat is handled by hand, irregularly.

What the paper contributes: *"Language Models Need Sleep: Learning to Self-Modify and Consolidate Memories"* (arXiv:2606.03979) frames consolidation as a **scheduled, offline, two-stage process** — distill fast/volatile memory into slow/stable memory, then prune the freed volatile capacity — with a concrete trigger schedule and a safety ordering (transfer before prune, freeze what you are not consolidating). The paper's actual engine is gradient-based and does not apply to text files, but its **schedule and safety ordering** map cleanly onto a markdown-file consolidation pass.

## 2. Design (adapted from the paper)

A standalone, idle-time pass that reads volatile memory and recent session signal, distills it into stable topic files, re-indexes, and archives what it pruned. The pass is an **LLM rewrite**, not a weight update — every paper mechanism below is adapted by swapping "backprop" for "the model reads and rewrites a file."

**Inputs (read-only at capture, no new instrumentation):**
- `MEMORY.md` (the volatile, always-loaded index — the "fast" store).
- The topic files under the memory dir (the "slow" store).
- Recent session signal already captured elsewhere: retro findings, user-correction/feedback memories (highest weight), cold-review findings, skill-drift findings, deny-streaks, no-commit terminations, phase-gate friction, eval regressions, handovers. These are the same channels the retro skill's orchestrator synthesis reads from (`skills/retro/SKILL.md:135`).
- `memory_audit.py`'s scanner output is a natural collector for the file side.

**Phases:**

| Phase | What it does | Paper concept adapted (arXiv:2606.03979) |
|---|---|---|
| **1. Replay / re-read** | Re-read the *raw* source files and recent session snapshots, not just yesterday's summary-of-summary. Guards against lossy-telephone drift. | Replay (§3.3, "teacher-generated data with masked prefixes … learns to complete them"). In files, replay = re-reading the source transcript instead of a summary-of-a-summary. |
| **2. Dedup / merge / cluster** | Group entries that share a root cause into one cluster. Identify near-duplicate feedback rules. | Multi-timescale hierarchy (§2.2): high-frequency volatile store vs. low-frequency stable store. The merge target is the stable store. |
| **3. Distill / compress (schema extraction)** | For each cluster, write a single imperative one-sentence rule. Turn N similar incidents into one rule — extract the schema, do not copy episodes verbatim. | Consolidation = distillation fast→stable (§3.1 NREM-inspired transfer; §3.3 Knowledge Seeding) + abstraction extraction (§2.2, lower-frequency layers store "abstract knowledge of their context"). |
| **4. (Optional) Dream / cross-link** | A low-temperature pass that deliberately cross-links *unrelated* entries to surface a latent shared root cause (e.g. "these 3 unrelated feedback items all stem from over-parallelization"). Keep only high-value links plus a few diverse ones; do not bloat. | Dreaming (§3.4): synthesize and recombine to surface hidden patterns; selection discipline = Top-k by importance + a few random for diversity. The randomness/novelty *idea* transfers; the MoE-router/gradient *mechanism* does not. |
| **5. Re-index** | Rewrite `MEMORY.md` so each surviving cluster has a ≤1-line index entry pointing at its topic file. Bring the index back under the load budget. | Consolidation timed to precede an overwrite (§3.2, "before each layer update, consolidation transfers knowledge to prevent knowledge override"). Overflow is the trigger. |
| **6. Decay / archive (prune)** | After a fact has a confirmed durable home in a topic file, remove its volatile index line to free the always-loaded budget. Archive (move to cold storage), never hard-delete; keep a restore-on-recurrence path. | Synaptic pruning (§3.3, "reset all the low-rank parameters previously added … making capacity available"). Pruning happens *after* transfer, never before. The index's existing "restore if a recurrence is observed" rule *is* re-potentiation. |

**Safety rails (non-negotiable):**
- **Transfer before prune.** Delete an index line only after the fact has a confirmed durable home in a topic file. Pruning before transfer is catastrophic forgetting (§3.3).
- **Freeze what you are not consolidating.** When rewriting a topic file, touch only the section being merged; leave settled sections byte-identical. No full-file regeneration (silent drop of old facts = catastrophic forgetting; §3.3, "freeze all parameters … only update the expanded").
- **Verify before durable write.** A consolidated rule is written only if it survives an adversarial check — "would this rule have prevented a real, cited mistake?" A side-observation is a lead, not a fact.
- **Never delete BINDING feedback without flagging.** BINDING / non-negotiable entries are never silently dropped or merged away. If a pass wants to retire one, it surfaces it for explicit user review instead of acting.
- **User-gated promotion of shared content.** Promoting any personal memory into a shared repo (skill / blueprint / public issue) stays user-gated and one-by-one — the consolidation pass may *propose*, never auto-publish.
- **Archive, don't destroy.** Pruned entries move to cold storage with a restore path; nothing is hard-deleted.

## Retro becomes emit-only (user-decided, same ticket)

Retro stops editing skill files directly. During the day it only (a) writes memories/structured signal and (b) files enforcement-gap issues. All consolidation, dedup, skill-file promotion and simplification move into this nightly dreaming pass. Exceptions that stay immediate: user-ordered fixes (normal coding dispatch) and same-day anti-vacuous evals for regressions (binding rule). Failure-mode rail: the dream loop needs a staleness alarm (doctor/statusline check: last successful dream > 48h ago) so memories never pile up unpromoted unnoticed. Scope: this issue also covers stripping the retro skill's § Fix Skills edit powers down to memory writes + issue filing (status: decided-by-user).

## 3. Trigger & placement

- **Cadence:** nightly / idle-time, decoupled from the live work loop. The paper makes consolidation periodic and frequency-tiered (§3.2, "memory consolidation happens only at {C⁽¹⁾b, …, C⁽ᵏ⁾b} steps") — cheap passes often, expensive passes rarely. Tier it: a cheap dedup-and-append every session; the expensive re-read + merge + re-index + archive on a coarser multiple (e.g. daily / every N sessions).
- **Threshold gate:** drain only when the count of unconsolidated observations is at or above a minimum batch (default 8), else no-op. This avoids waking the pass to do nothing.
- **Placement / extension point:** the reusable scaffolding is the loop's scanner-job pattern — a small dataclass with a cadence-elapsed check, a config on/off flag, and an in-flight lock. The closest existing analogue is the low-frequency, global-CORE scanners registered at `src/teatree/loop/tick_jobs.py:1268`:
  ```python
  jobs.extend(_ScannerJob(scanner=s, overlay="") for s in (_scanning_news_scanner(), _dogfood_smoke_scanner()) if s)
  ```
  with the builder pattern at `src/teatree/loop/tick_jobs.py:716` (`_scanning_news_scanner()`, returns `None` when disabled) and config keys at `src/teatree/config.py:336-339`. The architectural-review scanner (cadence 168h, `config.py:330-332`) is the closest "low-frequency, threshold-ish, not delivery-coupled" precedent.
  - **However**, the consolidation pass is heavier than a scanner tick and must not run on or re-arm the live 12-minute loop. Preferred placement is a **standalone low-frequency cron** (`t3 dream tick`, default ~04:00) that reuses the `ScannerJob` / cadence / config / in-flight-lock primitives but is driven by its own cron, not the work loop. Manual escape hatch: `t3 dream run [--since] [--dry-run]`.
- **Budget cap:** the expensive pass is wall-clock / token capped per run; if it exceeds budget it consolidates the highest-weight clusters first (user-correction memories outrank the rest) and defers the remainder to the next night. Off the critical path by construction.

## 4. Evals / acceptance

Each gate must **fail on a do-nothing (or a delete-only) consolidator** — that is what makes them anti-vacuous. Adapted from the paper's eval suite (§4.1), which directly measures "did consolidation preserve and improve knowledge."

| Gate | Procedure | Fails a no-op / delete-only pass because… | Paper source |
|---|---|---|---|
| **(a) Retention probe** | Before the pass, generate N QA pairs answerable from the pre-consolidation memory. After the pass, the consolidated memory must still answer all N. A dropped answer ⇒ lossy ⇒ reject. | A delete-only pass loses answers ⇒ fails. (A no-op passes this one, so it is not sufficient alone.) | Long-context retention — MK-NIAH, LongHealth, QASPER (§4.1). |
| **(b) Interference test** | After consolidating *this* session's facts, re-run the *previous* sessions' retention probes. Score must not regress. | A pass that overwrites unrelated stable sections regresses prior probes ⇒ fails. | Catastrophic-forgetting / class-incremental — CLINC, Banking, DBpedia (§4.1). |
| **(c) Consolidation-actually-happened** | The pass must reduce net size **or** increase the schema/cluster count (a merge occurred), AND every pruned index line must have a confirmed durable home. | A no-op changes nothing ⇒ size unchanged, schema count unchanged ⇒ fails. | Distillation ablation — the consolidation step is load-bearing (Table 1, §4.1). |
| **(d) Index back under budget** | After re-index, `MEMORY.md` must be under its line/size load warning (≤ the truncation threshold). | A no-op leaves `MEMORY.md` over budget ⇒ fails. | Overflow-as-trigger / consolidate-before-override (§3.2). |
| **(e) Monotonicity guard** | Running the pass twice over a stable corpus must not lower the retention-probe pass-rate (no summary-of-summary drift). | Over-compression on the second pass drops answers ⇒ fails. | Continual-learning monotonicity — Fig. 6, "performance improving monotonically with additional consolidation stages" (§4.1). |
| **(f) No-loss audit trail** | Every pruned/archived entry is recorded with its source and its durable destination; the archive is restorable. | A pass that hard-deletes leaves no trail ⇒ fails. | Synaptic pruning with re-potentiation (§3.3). |

Combined acceptance gate for one pass: **(a)** retention pass-rate ≥ pre-consolidation, **(b)** no regression on prior-session probes, **(c)** net size reduced or schema count increased, **(d)** `MEMORY.md` under its budget, **(e)** stable across repeated passes, **(f)** every pruned entry has a durable home and an audit-trail entry.

## 5. Non-goals (what does NOT transfer from the paper)

The paper's core engine is gradient-bound and is explicitly out of scope:
- **No fine-tuning / no weight updates.** Consolidation here is an LLM rewriting markdown files. The paper's SFT/LoRA, on-policy distillation objective, and RL imitation loop (§3.3/§3.4) have no file analogue — there are no weights to update.
- **No parameter expansion.** Adding low-rank experts before distilling (§3.2) is a weight-space capacity trick; not applicable.
- **No Knowledge Seeding / upward small→large distillation** (§3.3) — that transfers between two networks of different sizes; irrelevant to a single text store.
- **No Levenshtein / semantic RL reward** (Eq. 3, §3.3). We judge a rewrite's faithfulness with the retention probe (§4), not a reward-optimization loop.
- **No MoE router / random-expert mechanism** (§3.4). The novelty *idea* informs the optional dream/cross-link phase; the implementation does not carry over.
- **No new capture instrumentation.** The pass reads existing channels; it does not add new event logging.
- **No auto-publishing to shared repos.** Promotion of personal memory to skills/blueprint/public surfaces stays user-gated.
- **Not coupled to the live work loop.** It is a separate idle-time cron, never inline in a task turn and never re-arming the 12-minute tick.

## 6. Open questions & assumptions

- **[assumed]** Consolidation is a text rewrite over markdown files; no model fine-tuning is involved. (Direct consequence of a file-based memory.)
- **[assumed]** The expensive pass runs as a standalone low-frequency cron (`t3 dream tick`, ~04:00) reusing the `ScannerJob`/cadence/config/in-flight-lock primitives, not on the live loop. (Per the decoupling requirement; the scanner-job classes are scaffolding only.)
- **[assumed]** `DREAM_MIN_BATCH` default = 8 unconsolidated observations; cheap pass every session, expensive pass on a coarser multiple.
- **[assumed]** User-correction / feedback memories carry the highest weight when the per-run budget forces prioritization.
- **[open]** Output store: the related design favors a single DB model (`ConsolidatedMemory`, one row per cluster, status ladder CANDIDATE → VERIFIED → PROMOTED → SUPERSEDED → EXPIRED, idempotent `cluster_key`) rather than rewriting files in place. Decide: DB-backed consolidation vs. file-in-place rewrite vs. both (DB as the ledger, files as the rendered output). This changes phases 5–6 materially.
- **[open]** Where does the QA-probe corpus for the retention/interference gates come from, and is it persisted across runs (so monotonicity gate (e) can compare pass N+1 to pass N)?
- **[open]** Exact archive location and the restore-on-recurrence trigger — manual restore vs. automatic re-promote when a pruned rule's pattern recurs.
- **[open]** Should the optional dream/cross-link phase (4) ship in v1 or be deferred until the core distill/re-index/archive loop is proven? (It is the most speculative phase and the least load-bearing.)
- **[open]** Relationship to the dynamic-workflow mechanism: implement the pass as a dynamic workflow (mine → cluster → adversarially-verify → distill) vs. a dedicated CLI command. (The two are not mutually exclusive.)
- **[assumed]** This issue supersedes the narrower news-sourced candidate (MEMORY.md dedup/re-index only), which covers a strict subset of phases 2 and 5.

{"existing_issue": 1933, "approve_command": "python manage.py shell -c \"from teatree.core.models.pending_article_suggestion import PendingArticleSuggestion; PendingArticleSuggestion.objects.get(pk=20).approve(issue_url='<issue-url>')\""}

**Comment by souliane** [comment]
User-raised open question (2026-06-05): "maybe making teatree dream about the retro outputs only is not a good idea. should it dream directly from all conversation?"

Recommendation (adopt unless overridden): **dream from the raw session transcripts, with retro emissions as anchors — not from retro outputs alone.**

- **Primary replay source = raw transcripts** (`~/.claude/projects/*/<session>.jsonl`). Retro is in-context and lossy: it only emits what it noticed under context pressure, and a missed lesson never reaches the dream. Transcripts are ground truth and match the paper's episode-replay model (arXiv:2606.03979 replays raw episodes, not summaries).
- **Retro outputs become pre-tagged anchors**, not the sole input: memories, `t3-snapshot-*` files, task ledger, deferred questions, PR/issue trail all point the replay at the interesting regions of the transcript. Retro stays emit-only (as already decided), but dreaming no longer depends on retro's recall.
- **Cost bound:** map-reduce extraction over the since-last-dream transcript set with a cheap model tier for the map phase, a per-dream token budget, and the existing cadence/batch threshold (≥8) unchanged. Dreaming runs in idle windows, so latency is free; only tokens need capping.
- **Privacy:** transcripts contain user-personal content — the distill phase must keep the existing user-gated promotion rule for anything leaving local memory.

**Comment by souliane** [comment]
User addition (2026-06-05): dreaming should, if possible, identify questions/requests from the user that were ignored or left unanswered.

Concretely: when dreaming replays raw conversation transcripts (the direction chosen above), one extraction lane should mine every user message for questions/asks, cross-check each against the subsequent transcript + the answering-task ledger (did a threaded reply land? did the ask produce a task/PR?), and emit the unanswered ones as a surfaced backlog (DM digest + answering tasks). The existing answering-task scanner only sees messages it caught live — dreaming covers the misses: questions buried mid-message, multi-part asks where only part was handled, and FAILED answer tasks that were never retried.

Open question: dedupe against 'answered out-of-band' (user already got the answer elsewhere) — probably resolvable by checking for any agent reply in the message's thread.
